contracts/token/ERC721/ERC721Enumerable.sol

```diff
@@ -92,22 +91,26 @@ contract ERC721Enumerable is ERC165, ERC721, IERC721Enumerable {
     function _removeTokenFrom(address from, uint256 tokenId) internal {
         super._removeTokenFrom(from, tokenId);
 
-        // To prevent a gap in the array, we store the last token in the index of the token to delete, and
-        // then delete the last slot.
-        uint256 tokenIndex = _ownedTokensIndex[tokenId];
-        uint256 lastTokenIndex = _ownedTokens[from].length.sub(1);
-        uint256 lastToken = _ownedTokens[from][lastTokenIndex];
-
-        _ownedTokens[from][tokenIndex] = lastToken;
-        // This also deletes the contents at the last position of the array
-        _ownedTokens[from].length--;
-
-        // Note that this will handle single-element arrays. In that case, both tokenIndex and lastTokenIndex are going to
-        // be zero. Then we can make sure that we will remove tokenId from the ownedTokens list since we are first swapping
-        // the lastToken to the first position, and then dropping the element placed in the last position of the list
+        _removeTokenFromOwnerEnumeration(from, tokenId);
 
+        // Since the token is being destroyed, we also clear its index
+        // TODO(nventuro): 0 is still a valid index, so arguably this isnt really helpful, remove?
         _ownedTokensIndex[tokenId] = 0;
```

Comment:
You're only leaving this here to preserve the previous semantics but I _really_ don't like that `_removeTokenFrom` and `_removeTokenFromOwnerEnumeration` have such a subtle difference in behavior.

Reply:
Note that `_removeTokenFromOwnerEnumeration` is `private` and therefore nothing more than an implementation detail. I only added it to abstract away the repeated code in `_remove` and `_transfer`. If #1541 is accepted, then `_remove` will go the way of the dodo and the whole thing should end up looking cleaner: I kept that out of this PR on purpose.

Reply:
Wouldn't `_remove` be replaced by an internal `_burn` ? Won't we have the same problem there?

Reply:
Yes, but there's a specific reason for that (getting the refund - see #1541), and we'd only have one `_remove` _private_ function, not two. Again, its a case of trading _manageable_ complexity for efficiency.